File: Introduction to Robotics Course/Chapter12/Lesson6/Chapter12_Lesson6.py
# ROS2 Python node with reproducible randomness
import rclpy
from rclpy.node import Node
import numpy as np

class NoisyPublisher(Node):
    def __init__(self):
        super().__init__('noisy_pub')
        self.declare_parameter('seed', 0)
        seed = self.get_parameter('seed').value
        self.rng = np.random.default_rng(seed)  # deterministic PRG

        self.pub = self.create_publisher(Float32, 'noise', 10)
        self.timer = self.create_timer(0.1, self.step)

    def step(self):
        z = float(self.rng.normal(0.0, 1.0))
        msg = Float32()
        msg.data = z
        self.pub.publish(msg)
        self.get_logger().info(f"noise={z:.3f}")

def main():
    rclpy.init()
    node = NoisyPublisher()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()
      

File: Tutorials/Codes/Introduction_to_Robotics_Course/Chapter12/Lesson6/Chapter12_Lesson6.cpp
// ROS2 C++ example: seed passed as parameter
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/float32.hpp>
#include <random>

class NoisyPublisher : public rclcpp::Node {
public:
  NoisyPublisher() : Node("noisy_pub") {
    this->declare_parameter<int>("seed", 0);
    int seed = this->get_parameter("seed").as_int();

    rng_.seed(seed); // deterministic PRG

    pub_ = this->create_publisher<std_msgs::msg::Float32>("noise", 10);
    timer_ = this->create_wall_timer(
      std::chrono::milliseconds(100),
      std::bind(&NoisyPublisher::step, this));
  }
private:
  void step() {
    std::normal_distribution<float> dist(0.0f, 1.0f);
    float z = dist(rng_);
    std_msgs::msg::Float32 msg;
    msg.data = z;
    pub_->publish(msg);
  }
  std::mt19937 rng_;
  rclcpp::Publisher<std_msgs::msg::Float32>::SharedPtr pub_;
  rclcpp::TimerBase::SharedPtr timer_;
};
      
